src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder_ListPatterns.cs

```diff
@@ -55,17 +56,19 @@ private Tests MakeTestsAndBindingsForListPattern(BoundDagTemp input, BoundListPa
 
                         if (slice.Pattern is BoundPattern slicePattern)
                         {
-                            Debug.Assert(slice.IndexerAccess is not null);
-                            Debug.Assert(index <= 0);
-                            Debug.Assert(slice.ReceiverPlaceholder is not null);
-                            Debug.Assert(slice.ArgumentPlaceholder is not null);
+                            // We faced a slice pattern with a declaration pattern os the same narrowed type inside, e.g. `[<pattern>, ..var between, <pattern>]`
+                            // This pattern always evaluates to true, but causes slicing operation and even may allocate (e.g. when slicing an array or string)
+                            // Thus it makes sense to hold it and "evaluate" after all other patterns in the list matched
+                            if (slicePattern.Kind == BoundKind.DeclarationPattern && slicePattern.InputType.Equals(slicePattern.NarrowedType))
```

Comment:
Added `AllNullableIgnoreOptions` for now (this corresponds to `SymbolEqualityComparer.Default`), but I agree that it might be better to specify `AllIgnoreOptions`. So leaving this unresolved

Reply:
Are we settled on `AllNullableIgnoreOptions` or maybe a change is needed? It will be good if in case of change the observable effect can be tested, especially the runtime behavior

Reply:
Sounds good to me